doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
-   as a subgoal. This equality is automatically solved if it occurs among
-   the assumptions, or if its symmetric form occurs. It is equivalent to
-   :n:`cut @term = @term’; [intro H`:sub:`n` :n:`; rewrite <- H`:sub:`n` :n:`; clear H`:sub:`n`:n:`|| assumption || symmetry; try assumption]`.
+      This happens if :n:`@term` does not occur in, respectively, the named hypothesis or the goal.
 
-   .. exn:: Terms do not have convertible types.
-      :undocumented:
+   .. tacn:: erewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-   .. tacv:: replace @term with @term’ by @tactic
+      Works like :tacn:`rewrite`, but turns
+      unresolved bindings, if any, into existential variables instead of
+      failing. It has the same parameters as :tacn:`rewrite`.
 
-      This acts as :n:`replace @term with @term’` but applies :token:`tactic` to solve the generated
-      subgoal :n:`@term = @term’`.
+   .. flag:: Keyed Unification
 
-   .. tacv:: replace @term
+      Makes higher-order unification used by :tacn:`rewrite` rely on a set of keys to drive
+      unification.  The subterms, considered as rewriting candidates, must start with
+      the same key as the left- or right-hand side of the lemma given to rewrite, and the arguments
+      are then unified up to full reduction.
 
-      Replaces :n:`@term` with :n:`@term’` using the first assumption whose type has
-      the form :n:`@term = @term’` or :n:`@term’ = @term`.
+.. tacn:: rewrite * {? {| -> | <- } } @one_term {? in @ident } {? at @rewrite_occs } {? by @ltac_expr3 }
+          rewrite * {? {| -> | <- } } @one_term at @rewrite_occs in @ident {? by @ltac_expr3 }
+   :name: rewrite (star); _
+   :undocumented:
 
-   .. tacv:: replace -> @term
+.. tacn:: rewrite_db @ident {? in @ident }
+   :undocumented:
+
+.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }
+          replace {? {| -> | <- } } @one_term {? @occurrences }
+   :name: replace; _
 
-      Replaces :n:`@term` with :n:`@term’` using the first assumption whose type has
-      the form :n:`@term = @term’`
+   The first form replaces all free occurrences of :n:`@one_term__from`
+   in the current goal with :n:`@one_term__to` and generates an equality
+   :n:`@one_term__to = @one_term__from`
+   as a subgoal. (Note the generated equality is reversed with respect
+   to the order of the two terms in the tactic syntax; see
+   issue `#13480 <https://github.com/coq/coq/issues/13480>`_.)
+   This equality is automatically solved if it occurs among
+   the assumptions, or if its symmetric form occurs.
 
-   .. tacv:: replace <- @term
+   The second form, with `->` or no arrow, replaces :n:`@one_term`
+   with :n:`@one_term'` using
+   the first assumption whose type has the form :n:`@term = @term’`.
+   If `<-` is given, the tactic uses the first assumption with the reverse form,
+   i.e. :n:`@term’ = @term`.
```

Comment:
Where exactly do you get `term__to` from this form?

![image](https://user-images.githubusercontent.com/1253341/101730639-43e82400-3a6f-11eb-9e6e-fe8fe8d2d5c5.png)

Reply:
>Perhaps it's that `one_term` is itself an equality?

No. As explained in the text, you find the first assumption whose type is `term__from = term__to` (where `term__from` is the one you give as an argument to `replace` and `term__to` is just whatever is on the right-hand side of the equality). Said otherwise, what the tactic is looking for is rather `term__from = ?x` where whatever is unified with `?x` will replace `term__from` in the selected occurrences.

Reply:
OK, `replace`.  The updates are not so simple to apply, especially after I've been at it for a while.

Shouldn't it be like this, not quite what you said?

![image](https://user-images.githubusercontent.com/1253341/101856815-eb6e6080-3b1a-11eb-8b45-0eea0f53e550.png)

Reply:
LGTM.